Added BlueGiga USB dongle support for Bluetooth binding

kaikreuzer opened the pull request:
This is a follow-up on https://github.com/eclipse/smarthome/pull/4535, which had BlueGiga dongle support, while I had removed it from the merged https://github.com/eclipse/smarthome/pull/4997.

I have meanwhile cleaned up the code, fixed a few further bugs that I came across, added a test fragment, adapted the code to EPLv2, updated the license headers and added a README file.

Also-By: Chris Jackson <chris@cd-jackson.com>
Also-By: Vlad Kolotov <vkolotov@gmail.com>
Signed-off-by: Kai Kreuzer <kai@openhab.org>

htreu commented:
Unfortunately karaf feature verification fails:

```
missing requirement [org.eclipse.smarthome.binding.bluetooth.bluegiga/0.10.0.201807092041] osgi.wiring.package; filter:="(osgi.wiring.package=gnu.io)"
```

kaikreuzer commented:
This makes me think that it might be a good idea to also adapt the code to use the new ESH serial API. Will try to do so tonight!

kaikreuzer commented:
Done.

@maggu2810, when doing this change, I noticed that the ESH Serial API is lacking the methods `getName()`, `disableReceiveTimeout()`, `setFlowControlMode()`, `enableReceiveThreshold()` on `SerialPort` - the previous code used all of those on the RXTX version. They all seem to exist on `javax.comm.SerialPort` as well, so I am wondering, why you didn't add them to the ESH API?

maggu2810 commented:
I added the methods that are currently used.
I didn't see any reason to add all methods to our abstraction layer if there is no usage.

If we need that methods now it should be checked if every serial provider supports them and added.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.bluetooth.bluegiga/src/main/java/org/eclipse/smarthome/binding/bluetooth/bluegiga/handler/BlueGigaBridgeHandler.java:
```diff
@@ -0,0 +1,615 @@
+        command.setConnection(connectionHandle);
+        command.setStart(1);
+        command.setEnd(65535);
+        command.setUuid(UUID.fromString("00002800-0000-0000-0000-000000000000"));
```
This looks like a perfect candidate for a constant so this magic string gets a human readable meaning, doesn't it?

triller-telekom replied:
Did you see this comment?

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.bluetooth.bluegiga/src/main/java/org/eclipse/smarthome/binding/bluetooth/bluegiga/internal/BlueGigaEventListener.java:
```diff
@@ -0,0 +1,27 @@
+package org.eclipse.smarthome.binding.bluetooth.bluegiga.internal;
+
+/**
+ * 
```
Please add a description for this interface

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.bluetooth.bluegiga/src/main/java/org/eclipse/smarthome/binding/bluetooth/bluegiga/internal/BlueGigaPacket.java:
```diff
@@ -0,0 +1,23 @@
+ * @author Chris Jackson
+ *
+ */
+public class BlueGigaPacket {
```
Given the size of this PR and having a packet format specific to this dongle, I am wondering whether each bluetooth dongle sends different packets? If not, can't we move these definitions into the common bluetooth binding and export them there for all dongles to use?

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.bluetooth.bluegiga/src/main/java/org/eclipse/smarthome/binding/bluetooth/bluegiga/internal/BlueGigaSerialHandler.java:
```diff
@@ -0,0 +1,505 @@
+    private boolean notifyTransactionComplete(final BlueGigaResponse response) {
+        boolean processed = false;
+
+        // logger.debug("NODE {}: notifyTransactionResponse {}",
```
left over commented out code

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.bluetooth.bluegiga/src/main/java/org/eclipse/smarthome/binding/bluetooth/bluegiga/internal/BlueGigaSerialHandler.java:
```diff
@@ -0,0 +1,505 @@
+        }
+    }
+
+    // TODO: Add a timeout mechanism
```
Seeing the `TRANSACTION_TIMEOUT_PERIOD` parameter I guess this comment is now obsolete?

htreu commented on extensions/binding/org.eclipse.smarthome.binding.bluetooth.bluegiga/src/main/java/org/eclipse/smarthome/binding/bluetooth/bluegiga/internal/BlueGigaResponsePackets.java:
```diff
@@ -0,0 +1,204 @@
+/**
+ * Helper class to create BlueGiga BLE Response and Event packets (ie packets that we will receive).
+ * <p>
+ * Note that this code is autogenerated. Manual changes may be overwritten.
```
I noticed several classes to have this note. Why do we check in autogenerated code? What was the source of this files? Wouldn't it be more future prove to add the original source and the generator to the project?

maggu2810 commented on extensions/binding/org.eclipse.smarthome.binding.bluetooth.bluegiga/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,30 @@
+Service-Component: OSGI-INF/*.xml
+Export-Package: org.eclipse.smarthome.binding.bluetooth.bluegiga,
+ org.eclipse.smarthome.binding.bluetooth.bluegiga.handler
+Bundle-ActivationPolicy: lazy
```
Remove that line and add `Automatic-Module-Name`

kaikreuzer commented:
Thanks @maggu2810, @triller-telekom & @htreu, I have incorporated all your review comments.
Wrt the generated code: I believe that this was mainly a one-time generation to avoid a lot of typing; I do not see that this is required as a regular step for maintaining this code. Nonetheless, I kept the warning in the JavaDoc in place as touching code that has once been generated usually adheres to a strict structure where manual changes are likely to break things unintentionally.

@triller-telekom I do not know whether this code is only applicable for BlueGiga or also to other dongles / BT support. But as long as nobody expresses the need to reuse it at some other place, I very much prefer to keep it internal here. I myself have never missed it in the generic bluetooth support.

@maggu2810 I have added the additional methods in the ESH Serial API within this PR, I hope this is ok.

In general, I'd like to mention that this code might not in all places be fully stable, tested and production-ready - this is why I kept is so long in my own fork. But as it is immensely useful when developing bluetooth stuff on a non-Linux machine, I think it should be made available in ESH to all developers. In a typical production scenario, almost everyone will probably rather use Linux with BlueZ.

I've kept the PR as a single commit. If you do not find any severe issues with it anymore, it would be great if a CQ could be filed for it.

triller-telekom commented on extensions/binding/org.eclipse.smarthome.binding.bluetooth.bluegiga/cfg/ble.bluegiga.cfg:
```diff
@@ -0,0 +1 @@
+# Ble.BlueGiga Binding Default Configuration
```
Is there a reason for not removing this empty configuration file, or did you just overlook it?

maggu2810 commented:
> @maggu2810 I have added the additional methods in the ESH Serial API within this PR, I hope this is ok.

As there also exists #5560 that adds another serial implementation have you already checked if this implementation supports the implementation of the new interface methods, too?
I assume that the first merge breaks the second one.
If #5560 is merged, you have to add the interface implementation to the new serial implementation.
If this one is merged, @msteigenberger has to add the interface implementation to fit to the new serial interface.

maggu2810 commented:
Didn't we agree that new classes / interfaces should already contain the `@NonNullByDefault` annotation?

kaikreuzer commented:
> You seem to have forgotten one emtpy config file

Indeed, that got lost in the commit, sorry. Not it is removed.

> there is a magic UUID as a string in the code

The reason I left it there is that it is a magic string for me, too; I do not know, what a good constant name would be for it.

> Didn't we agree that new classes / interfaces should already contain the @NonNullByDefault annotation?

I do not remember such a strict decision. If so, this would be mainly for framework interfaces, but I do not think we enforce it for any internal class of any extension. We only expect that no `@NonNull` annotations are used and this is what I have already fixed after @triller-telekom's remark.
Note that this code is also already a year old and I didn't write it - I merely think it should be added to the repo, but I don't want to take the effort to do huge refactorings to it.

> have you already checked if this implementation supports the implementation of the new interface methods, too?

Good point. I just checked: Yes, the TelnetSerialPort also has those methods, so it should be easy to adapt #5560 accordingly. To avoid dependencies between this PR and #5560, I have now created the separate PR https://github.com/eclipse/smarthome/pull/5907, which should be merged before any of the other two. This PR here will now hence fail to compile until #5907 is merged - but it imho should not prevent us from already creating a CQ for it.

cdjackson commented:
I just noticed that this PR copies the source from https://github.com/zsmartsystems/com.zsmartsystems.bluetooth.bluegiga. Can you add a note to this effect in the NOTICE file please?

kaikreuzer commented:
The code actually originates from extracting the sources that were included in the zip in https://github.com/eclipse/smarthome/pull/3633, which had then been modified in https://github.com/eclipse/smarthome/pull/4535 to match the ESH bluetooth API.
You are right, this effectively means they are a modified version of the code at https://github.com/zsmartsystems/com.zsmartsystems.bluetooth.bluegiga.

The NOTICE file is usually used for referencing third party code that is included as a binary in the distribution (so that it is clear where to find the source code). But if you wish to also add some note wrt zsmartsystems, please suggest a text and I will add it.

kaikreuzer commented:
@maggu2810 & @htreu: I have fixed the travis build issue, so imho this is ready for a CQ - any further changes can be done through separate follow-PRs.

htreu commented:
FTR: CQ16983 created.

maggu2810 commented:
CQ16983 has been approved.

cdjackson commented:
@kaikreuzer sorry - I didn't see your response to the above. Note that if you're using the original code from 18 months ago, then there were a number of bug fixes since then.

Since this has been merged now I will create a separate PR to reference the library if that's ok?
